src/main/scala/com/gu/Box.scala

```diff
@@ -32,7 +33,8 @@ private class AtomicRefBox[T](t: T) extends Box[T] {
   def send(f: T => T): Unit = ref.updateAndGet(t => f(t))
 
   def alter(t: T): Future[T] = Future.successful(ref.updateAndGet(_ => t))
-  def alter(f: T => T): Future[T] = Future.successful(ref.updateAndGet(t => f(t)))
+  def alter(f: T => Try[T])(implicit ec: ExecutionContext): Future[T] = 
+    Future.fromTry(f(t)).map(t => ref.updateAndGet(_ => t))
```

Comment:
I think if you pattern match on the result of f(t), you won't have to map on the "fake" future, and therefore won't need the execution context.

Something like that?

```
f(t) match {
  case Success(value) => Future.successful(ref.updateAndGet(_ => t))
  case Failure(error) => Future.fail(error)
}
```

Reply:
Yes, you're right!